chromadb/auth/basic_authn/__init__.py

```diff
@@ -0,0 +1,130 @@
+    """
+    Client auth provider for basic auth. The credentials are passed as a
+    base64-encoded string in the Authorization header prepended with "Basic ".
+    """
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self._settings = system.settings
+        system.settings.require("chroma_client_auth_credentials")
+        self._creds = SecretStr(
+            str(system.settings.chroma_client_auth_credentials)
+        )
+
+    @override
+    def authenticate(self) -> ClientAuthHeaders:
+        # We do this goofy encode-decode dance to make the type checker happy.
+        encoded = base64.b64encode(
+            f"{self._creds.get_secret_value()}".encode("utf-8")
+        ).decode(
+            "utf-8"
+        )
+        return {
+            "Authorization": SecretStr(
+                f"Basic {encoded}"
+            ),
+        }
+
+
+class BasicAuthenticationServerProvider(ServerAuthenticationProvider):
+    """
+    Server auth provider for basic auth. The credentials are read from
+    `chroma_server_authn_credentials_file` and each line must be in the format
+    <username>:<bcrypt passwd>.
+
+    Expects tokens to be passed as a base64-encoded string in the Authorization
+    header prepended with "Basic".
+    """
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self._settings = system.settings
+
+        try:
+            # We need this to check passwords
+            self.bc = importlib.import_module("bcrypt")
+        except ImportError:
+            raise ValueError(
+                "The bcrypt python package is not installed. "
+                "Please install it with `pip install bcrypt`"
+            )
+
+        self._creds: Dict[str, SecretStr] = {}
+        creds = self.read_creds_or_creds_file()
+
+        for line in creds:
+            if not line.strip():
+                continue
+            _raw_creds = [v for v in line.strip().split(":")]
+            if (_raw_creds and _raw_creds[0] and
+                    len(_raw_creds) != 2 or not all(_raw_creds)):
+                raise ValueError(
+                    f"Invalid htpasswd credentials found: {_raw_creds}. "
+                    "Lines must be exactly <username>:<bcrypt passwd>."
+                )
+            username = _raw_creds[0]
+            password = _raw_creds[1]
+            if username in self._creds:
+                raise ValueError(
+                    "Duplicate username found in "
+                    "[chroma_server_authn_credentials]. "
+                    "Usernames must be unique."
+                )
+            self._creds[username] = SecretStr(password)
+
+    @trace_method("BasicAuthenticationServerProvider.authenticate",
+                  OpenTelemetryGranularity.ALL)
+    @override
+    def authenticate_or_raise(
+        self, headers: Headers
+    ) -> UserIdentity:
+        try:
+            _auth_header = headers["Authorization"]
```

Comment:
what happens if someone sends malformed headers that don't match these assumptions?

Reply:
We raise a `KeyError` which is caught on line 116, then we raise a 403 Forbidden on line 121